### PR TITLE
WIP: Lazy work

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs [hunk 1]
```diff
@@ -23,6 +24,30 @@ public static AsyncLazy<T> Create<T>(T value)
         => new(value);
 }
 
+internal sealed class AsyncLazy<T> : AsyncLazy<
+    T,
+    (Func<CancellationToken, Task<T>> asynchronousComputeFunction, Func<CancellationToken, T>? synchronousComputeFunction)>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs:29]: @ToddGrun here's how to pass things along without allocs.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs [hunk 2]
```diff
@@ -23,6 +24,30 @@ public static AsyncLazy<T> Create<T>(T value)
...
+        : this(asynchronousComputeFunction, synchronousComputeFunction: null)
+    {
+    }
+
+    public AsyncLazy(
+        Func<CancellationToken, Task<T>> asynchronousComputeFunction,
+        Func<CancellationToken, T>? synchronousComputeFunction)
+        : base(
+            asynchronousComputeFunction: static (t, c) => t.asynchronousComputeFunction(c),
+            synchronousComputeFunction: synchronousComputeFunction is null ? null : static (t, c) => t.synchronousComputeFunction!(c),
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs:45]: and here we ensure that the lambdas here are static.

Comment (ToddGrun) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs:45]: Compiler isn't smart enough to figure out the bang isn't necessary on "t.synchronousComputeFunction!(c)"?

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs:45]: It is not.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs [hunk 3]
```diff
@@ -23,6 +24,30 @@
...
 
+internal sealed class AsyncLazy<T> : AsyncLazy<
+    T,
+    (Func<CancellationToken, Task<T>> asynchronousComputeFunction, Func<CancellationToken, T>? synchronousComputeFunction)>
+{
+    public AsyncLazy(T value) : base(value)
+    {
+    }
+
+    public AsyncLazy(Func<CancellationToken, Task<T>> asynchronousComputeFunction)
```
Comment (ToddGrun) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs:35]: Something feels a bit off, for example I added the following method to AsyncLazy:

    public static AsyncLazy<T, TData> Create<T, TData>(Func<TData, CancellationToken, Task<T>> asynchronousComputeFunction, TData data)
        => new(asynchronousComputeFunction, data);

and then I switched SolutionCompilationState._lazyChecksums over by making the following changes. It felt wrong to include the data type in the declaration

``` 
// original code
private AsyncLazy<SolutionCompilationStateChecksums> _lazyChecksums;

private SolutionCompilationState(...)
{
    _lazyChecksums = AsyncLazy.Create(c => ComputeChecksumsAsync(projectId: null, c));
}
```

and I needed to specify the data type in the member declaration and wasn't able to use the Create

``` 
// new code
private readonly AsyncLazy<SolutionCompilationStateChecksums, SolutionCompilationState> _lazyChecksums;

private SolutionCompilationState(...)
{
    _lazyChecksums = AsyncLazy.Create(static (self, c) => self.ComputeChecksumsAsync(projectId: null, c), data: this);
}
```

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs:35]: is it any better if you reverse the order of the arguments?

**Comment by CyrusNajmabadi** [comment]
Closing out as this was completed by todd.

**Pull request closed without merging**